src/download.rs

```diff
@@ -0,0 +1,107 @@
+//! Represents a file to be downloaded.
+
+use crate::Error;
+use std::convert::TryFrom;
+use url::Url;
+
+/// Represents a file to be downloaded.
+#[derive(Debug, Clone)]
+pub struct Download {
+    /// URL of the file to download.
+    pub url: Url,
+    /// Filename used to save the file on disk.
+    pub filename: String,
+}
+
+impl Download {
+    /// Creates a new [`Download`].
+    ///
+    /// When using the [`Download::try_from`] method, the filename is
+    /// automatically extracted from the URL.
+    ///
+    /// ## Example
+    ///
+    /// The following calls are equivalent, minus some extra URL validations
+    /// performed by `try_from`:
+    ///
+    /// ```no_run
+    /// # use color_eyre::{eyre::Report, Result};
+    /// use trauma::download::Download;
+    /// use url::Url;
+    ///
+    /// # fn main() -> Result<(), Report> {
+    /// Download::try_from("https://example.com/file-0.1.2.zip")?;
+    /// Download::new(&Url::parse("https://example.com/file-0.1.2.zip")?, "file-0.1.2.zip");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new(url: &Url, filename: &str) -> Self {
+        Self {
+            url: url.clone(),
+            filename: String::from(filename),
+        }
+    }
+}
+
+impl TryFrom<&Url> for Download {
+    type Error = crate::Error;
+
+    fn try_from(value: &Url) -> Result<Self, Self::Error> {
+        let path_segments = value.path_segments();
+        if let Some(u) = path_segments {
```

Comment:
[`Option::map`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.map) and [`Option::transpose`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.transpose) can be used here:
```rs
path_segments
  .map(|u| u.last())
  .map(String::from)
  .map(|filename| Download{ url: value.clone(), filename})
  .transpose()
  // .map_err(|e| e) handle error here
```

Reply:
I could not get it to work like you proposed but I ended up with this solution:

```rust
value
            .path_segments()
            .ok_or(Error::InvalidUrl(format!("")))?
            .last()
            .map(String::from)
            .map(|filename| Download {
                url: value.clone(),
                filename,
            })
            .ok_or(Error::InvalidUrl(format!("")))
```

Not sure if it is the most idiomatic, but it reads OK and works 😅